src/DynamoDBMessageRepository/DynamoDBMessageRepository.php

```diff
@@ -0,0 +1,140 @@
+    {
+        $this->client = $client;
+        $this->serializer = $serializer;
+        $this->tableName = $tableName;
+    }
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        $items = [];
+        $marshaler = new Marshaler();
+
+        foreach ($messages as $message) {
+            $payload = $this->serializer->serializeMessage($message);
+            $event = $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
+            $aggregateRootId = $payload['headers'][Header::AGGREGATE_ROOT_ID] ?? null;
+            $eventType = $payload['headers'][Header::EVENT_TYPE] ?? null;
+            $aggregateRootVersion = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
+            $timeOfRecording = $payload['headers'][Header::TIME_OF_RECORDING];
+
+            $item = [
+                'event' => $marshaler->marshalValue($event),
+                'aggregateRootId' => $marshaler->marshalValue($aggregateRootId),
+                'eventType' => $marshaler->marshalValue($eventType),
+                'aggregateRootVersion' => $marshaler->marshalValue($aggregateRootVersion),
+                'timeOfRecording' => $marshaler->marshalValue($timeOfRecording),
+                'payload' => $marshaler->marshalItem(['payload' => $payload])['payload']
+            ];
+
+            $items[] = [
+                'PutRequest' => [
+                    'Item' => $item
+                ]
+            ];
+        }
+
+        $batchRequest = ['RequestItems' => [ $this->tableName => $items ]];
+
+        try {
+            $this->client->batchWriteItem($batchRequest);
+        } catch (Throwable $exception) {
+            throw UnableToPersistMessages::dueTo('', $exception);
+        }
+
+    }
+
+    public function retrieveAll(AggregateRootId $id): Generator
+    {
+        $query = [
+            'TableName' => $this->tableName,
+            'KeyConditionExpression' => 'aggregateRootId = :v1',
+            'ExpressionAttributeValues' => [
+                ':v1' => ['S' => $id->toString()],
+            ]
+        ];
+
+        $result = $this->client->getPaginator('Query', $query);
+
+        try {
+            $result->valid();
+            return $this->yieldMessagesForResult($result);
+        } catch (Throwable $exception) {
+            throw UnableToRetrieveMessages::dueTo('', $exception);
+        }
+    }
+
+    /** @psalm-return Generator<Message> */
+    public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootVersion): Generator
+    {
+        $query = [
+            'TableName' => $this->tableName,
+            'KeyConditionExpression' => 'aggregateRootId = :v1 AND aggregateRootVersion > :v2',
+            'ExpressionAttributeValues' => [
+                ':v1' => ['S' => $id->toString()],
+                ':v2' => ['N' => $aggregateRootVersion]
+            ]
```

Comment:
This is in no way wrong, but I'd generally try to stick with named paraters, instead of numbered, easier to read/maintain (but feel free to ignore this suggestion).

```suggestion
            'KeyConditionExpression' => 'aggregateRootId = :aggregateRootId AND aggregateRootVersion > :aggregateRootVersion',
            'ExpressionAttributeValues' => [
                ':aggregateRootId' => ['S' => $id->toString()],
                ':aggregateRootVersion' => ['N' => $aggregateRootVersion]
            ]
```

Also, why not use the marshaller here?